addons/mail/static/src/messaging/entity/messaging/messaging.js

```diff
@@ -53,24 +53,24 @@ function MessagingFactory({ Entity }) {
         async redirect({ id, model }) {
             if (model === 'mail.channel') {
                 const channel = this.env.entities.Thread.channelFromId(id);
-                if (!channel) {
+                if (!channel || !channel.isPinned) {
                     this.env.entities.Thread.joinChannel(id, { autoselect: true });
                     return;
                 }
                 channel.open();
             } else if (model === 'res.partner') {
-                if (id === this.currentPartner) {
+                if (id === this.currentPartner.id) {
```

Comment:
Shouldn't we add a test covering that?

Reply:
You mean, the redirect method ? Should be done in this branch you think ?

Reply:
If there was a change needed here, that means there was a bug, so there should be a test covering it, because that seems tricky. I would tend to agree with you, it would be better to add that test in another branch (and the fix too).

Reply:
Made change because this test was not working as expected: _"click on partner follower details"_

Test asserts it opens form view of partner, and `messaging:redirect()` does that in case the redirect happens on current partner... Yeah, this method has been designed only with discuss app and chat window in mind, not with chatter, that's why it opens chat window with other partners instead.

I think `redirect()` should receive some context, so that from a click on partner follower it always open form view. As for channels... maybe also open the form view? I dunno ¯\_(ツ)_/¯

Reply:
If you dunno then I dunno what should be changed here :<

Reply:
I think I prefer open form view on click of channel follower, for consistency with partner followers and I also think joining a channel is an undesired commitment in this case.

Reply:
Yeah but we should do in another function or add some kind of parameter to `redirect` function to allow the openDocument to be done in that case instead of joinChannel ?

Reply:
This is tricky, because we have to be clear on what is meant by "redirect".

After some thinking, I think that redirect means "open something" with provided id and model.
Usually in Odoo it means open form view.
The particularity of discuss is that `mail.channel` is strange model, and so redirect instead should open a thread viewer (Discuss or Chat window depending on whether discuss app is open or not).
The special redirect on `res.partner` is there to handle `mail.channel` of chat.
That one is questionable, because contrary to `mail.channel` being a special model, `res.partner` is not, so that's ballsy to assume such a redirect always mean open chat...

All of this just to say that something is weird with redirect...
Seems reasonable to sometimes expect a redirect as "open record" and sometimes as "open thread"...
It is as if a click on such redirection link should prompt either choice to user and/or link itself should clearly tell whether this redirection is on record or the thread...

For the scope of this PR, and for the moment, let's just add a parameter to redirect so that it simply open document.